src/llfs/ioring_log_device2.test.cpp

```diff
@@ -0,0 +1,594 @@
+      //
+      if (!this->appended_slots.empty()) {
+        ASSERT_GE(this->appended_slots.front().lower_bound, slot_range.upper_bound);
+      }
+    }
+
+    /** \brief The simulation entry point.
+     *
+     * Consists of two phases:
+     *   1. append phase - append slots until we reach kNumSlots or an injected error stops us from
+     *      making progress
+     *   2. verify phase - recover the log and check flush/trim pos and slot data
+     *
+     * If we run out of space in (1), we trim slots one at a time until the desired amount of space
+     * has been freed up.
+     */
+    void run()
+    {
+      ASSERT_NO_FATAL_FAILURE(this->initialize_log_storage());
+
+      this->sim.run_main_task([this] {
+        {
+          ASSERT_NO_FATAL_FAILURE(this->recover_log_device());
+          auto on_scope_exit = batt::finally([&] {
+            this->shutdown_log_device();
+          });
+
+          this->sim.set_inject_failures_mode((seed % 2) == 1);
+
+          for (usize i = 0; i < kNumSlots; ++i) {
+            LLFS_VLOG(1) << "Writing slot " << i << " of " << kNumSlots << ";"
+                         << BATT_INSPECT(this->total_bytes_appended) << BATT_INSPECT(seed);
+
+            if (!this->append_one_slot()) {
+              break;
+            }
+          }
+        }
+
+        this->sim.crash_and_recover();
+        this->sim.set_inject_failures_mode(false);
+
+        {
+          ASSERT_NO_FATAL_FAILURE(this->recover_log_device());
+          auto on_scope_exit = batt::finally([&] {
+            this->shutdown_log_device();
+          });
+
+          llfs::SlotRange slot_range = this->log_device->slot_range(llfs::LogReadMode::kDurable);
+
+          ASSERT_NO_FATAL_FAILURE(this->verify_recovered_slot_range(slot_range));
+          ASSERT_NO_FATAL_FAILURE(this->resolve_maybe_trimmed_slots(slot_range));
+          ASSERT_NO_FATAL_FAILURE(this->verify_recovered_slots(slot_range));
+        }
+      });
+    }
+
+  };  // struct Scenario
+
+};  // class IoringLogDevice2SimTest
+
+//=#=#==#==#===============+=+=+=+=++=++++++++++++++-++-+--+-+----+---------------
+
+TEST_F(IoringLogDevice2SimTest, Simulation)
+{
+  const usize kNumThreads = std::thread::hardware_concurrency();
+  const usize kUpdateInterval = kNumSeeds / 20;
+
+  LLFS_LOG_INFO() << BATT_INSPECT(kNumSeeds);
+
+  std::vector<std::thread> threads;
+  usize start_seed = 0;
+  usize seeds_remaining = kNumSeeds;
+  for (usize i = 0; i < kNumThreads; ++i) {
+    usize n_seeds = seeds_remaining / (kNumThreads - i);
+    usize end_seed = start_seed + n_seeds;
+    threads.emplace_back([&, start_seed, end_seed] {
+      for (usize seed = start_seed; seed < end_seed; seed += 1) {
+        LLFS_LOG_INFO_EVERY_N(kUpdateInterval)
+            << "progress=" << (seed - start_seed) * 100 / (end_seed - start_seed) << "%";
+
+        Scenario scenario{seed};
+        ASSERT_NO_FATAL_FAILURE(scenario.run());
+      }
+    });
+    start_seed += n_seeds;
```

Comment:
It looks like the  `start_seed`s used will always be the same in this test case. Consider adding a test case where we use a random  start_seed every time. I think doing so may increase our chances of getting lucky and finding a new interleaving that breaks something.

Reply:
I'll make this controllable by env var, but I don't want to automatically randomize it each time, so that tests will be deterministic.

Reply:
sgtm.